### ISSUE TITLE
german keyboard on MacOS

### ISSUE THREAD

**Issue by hjanuschka** [equalsraf/neovim-qt#833]
damn, the `--server` is a gamechanger!!!! huge props!

but, i have problems to enter some characters on my german macos qwertz keyboard, like `{` or `}` 

any ideas?

**Comment by jgehrig** [comment]
Sure! We can get this resolved...

First thing to try is using a build from the latest commit:
```
$ brew tap equalsraf/neovim-qt
$ brew install --HEAD neovim-qt
```

I think there is a [typo](https://github.com/equalsraf/neovim-qt#macos) in my recent README update. Please let me know if the `--HEAD` above works for you.

You can check for the latest version with `--version`:
```
$ nvim-qt --version
NVIM-QT v0.2.17.9999
Build type: Release
Compilation:-march=native -O2 -pipe -Wall -Wextra -Wno-unused-parameter -Wunused-variable -std=c++11
Qt Version: 5.15.2
Environment: 
  nvim: nvim
  args: --cmd let &rtp.=',/usr/bin/../share/nvim-qt/runtime' --cmd set termguicolors
  runtime: /usr/bin/../share/nvim-qt/runtime
```

If your `{` key still doesn't work, I will need some output from a debug build to diagnose the issue:

https://github.com/equalsraf/neovim-qt/blob/621495a49df778d5f3e2625fb16d59c3483026e8/src/gui/shell.cpp#L1242-L1244

1. Un-comment the lines above
2. Build `neovim-qt` from source [MacOS Bulid Instructions](https://github.com/equalsraf/neovim-qt/wiki/Build-Instructions#macos)
2. Make sure to use `-DCMAKE_BUILD_TYPE=Debug`
3. Press `{` and `}` using your keyboard layout.
4. Paste the debug output here.

I'm working to slowly improve the input layer and add test coverage. I will use your debug output to write tests, and then fix the issue (if necessary).

Thanks!

**Comment by hjanuschka** [comment]
this is what happens if i press `{`   - like on german keyboard -> OPTION+8


```

QKeyEvent ev: QKeyEvent(KeyPress, Key_Alt, AltModifier)
   ""
QKeyEvent ev: QKeyEvent(KeyPress, Key_8, AltModifier, text="{")
   "<A-{>"
QKeyEvent ev: QKeyEvent(KeyPress, Key_Control, ControlModifier)


```


this is what happens on the `}` - OPTION+9

```
QKeyEvent ev: QKeyEvent(KeyPress, Key_Alt, AltModifier)
   ""
QKeyEvent ev: QKeyEvent(KeyPress, Key_9, AltModifier, text="}")
   "<A-}>"
QKeyEvent ev: QKeyEvent(KeyPress, Key_Control, ControlModifier)
```

**Comment by jgehrig** [comment]
Thanks for the details! Logs and bug reports like this are essential to improving the input layer...

We are not removing the Alt Modifier when we should. I will put together a fix and Pull Request.

Just to clarify, the log output is from the latest commit to `master`, right?

**Comment by hjanuschka** [comment]
yes its from latest master

fresh clone - patch n build!

**Comment by jgehrig** [comment]
Thanks! Pull Request posted...

Do you know how the German layout handles `{}` on other platforms?

It would be nice to add test coverage and validate on Linux/Windows too.

From quick test on Linux of `Alt + 9` is `QKeyEvent ev: QKeyEvent(KeyPress, Key_9, AltModifier, text="9")`. No `{}`.

**Comment by hjanuschka** [comment]
can confirm that `{}` now works. just figured out another essential one `|` 



```
QKeyEvent ev: QKeyEvent(KeyPress, Key_Alt, AltModifier)
   ""
QKeyEvent ev: QKeyEvent(KeyPress, Key_7, AltModifier, text="|")
   "<A-|>"
QKeyEvent ev: QKeyEvent(KeyPress, Key_Control, ControlModifier)
   ""

```


its OPTION+7

**Comment by hjanuschka** [comment]
gave it a try a couple of work hours! so far amazing, after setting font and stuff :)


but have another (i guess this time macOS related) issue, when trying to paste - COMMAND+P from - a different window, into nvim (insert mode)



```
QKeyEvent ev: QKeyEvent(KeyPress, Key_Control, ControlModifier)
   ""
QKeyEvent ev: QKeyEvent(KeyPress, Key_V, ControlModifier)
   "<D-v>"
QKeyEvent ev: QKeyEvent(KeyPress, Key_Control, ControlModifier)
   ""
```

**Comment by jgehrig** [comment]
Regarding `|`. Good to know, the detection logic for my fix should be made more generic.

Perhaps we can detected based on: AltModifier, Ascii Range, and Key does not match Text?

> but have another (i guess this time macOS related) issue, when trying to paste - COMMAND+P from - a different window, into nvim (insert mode)

I don't see anything wrong with the  input here. Commands `CTRL + V` and `CTRL + P` have no special meaning in vim.

You'll have to add a map to `init.vim` or `vimrc`:
```
cmap <???> <C-R>+
imap <???> <C-R>+
```

The `???` should be replaced with whatever key sequence you prefer: `S-Insert`, `C-V`, `C-P`.

> i am so amazed that this is a high-potential that i can get rid of vscode - and return to vim with solid remote caps.
going to dig into the code, and hopefully i ll be able to help/contribute.

Yes, the architecture of Neovim is quite clever. The separation between backend editor and GUI works well, enabling features like `--server` and `nvr`.  

This would be great! Any help would be appreciated...

**Comment by hjanuschka** [comment]
ohhh thx, for the cmap/imap, forgot about it, as it worked out of the box in terminal guess my terminal does that for me. adding the remapping fixes copy & paste in neovim-qt

**Comment by jgehrig** [comment]
No worries!

One more question:
Do you know how these keys behave with the German layout on Windows/Linux?

Ideally, I'd like to add test coverage for those as well...

**Comment by hjanuschka** [comment]
see the PR - i have changed it to be a bit more generic.
i have no linux booted here, but keys are slightly diff. on PC keyboards


like:


PC: `{`  -> ALTGR+7
MAC: `{` -> OPTION+7 (with pc keyboard attached to mac - it is ALT (not GR))

**Comment by jgehrig** [comment]
Closing, merged in #841.
